src/ir/auth_logic/lowering_ast_datalog.h

```diff
@@ -0,0 +1,319 @@
+    return Predicate(new_name, new_args, sign_copy);
+  }
+
+  // This function is an abbreviation for `PushPrin` where:
+  // - a modifier is added
+  // - just one new principal is added as an argument.
+  // This is a common case in this translation because it is used for
+  // `x says blah(args)` and `x canActAs y` and other constructions involving a
+  // principal name.
+  Predicate PushPrincipal(std::string modifier, const Principal& principal,
+                          const Predicate& predicate) {
+    return PushOntoPredicate(std::move(modifier), {principal.name()},
+                             predicate);
+  }
+
+  Predicate AttributeToDLIR(const Attribute& attribute) {
+    // If attribute is `X pred(args...)` the following predicate is
+    // `pred(X, args...)`
+    return PushPrincipal(std::string(""), attribute.principal(),
+                         attribute.predicate());
+  }
+
+  // Translating an attribute results in two objects:
+  // - A predicate, this is the main result of the translation
+  // - An additional rule that allows attributes to be passed with canActAs
+  // This function is for generating the additional rule.
+  // This takes both a speaker and the actual attribute as argument because
+  // the speaker is needed to generate the additional rule. Note that all base
+  // facts have a speaker either explicitly or implicitly:
+  // - If it appears on the LHS of an assertion, it explicitly has a speaker
+  // - If it appears on the RHS of an assertion, it behaves semantically
+  // like it has the same speaker as the head of the assertion.
+  DLIRAssertion SpokenAttributeToDLIR(Principal speaker, Attribute attribute);
+
+  Predicate CanActAsToDLIR(const CanActAs& can_act_as) {
+    return Predicate("canActAs",
+                     {can_act_as.left_principal().name(),
+                      can_act_as.right_principal().name()},
+                     kPositive);
+  }
+
+  // In the same way that attributes are passed around with "CanActAs", so
+  // are other "CanActAs" facts. To implement this, the translation of
+  // CanActAs also results in both a predicate and an extra rule that passes
+  // these facts around. This function is for generating the extra rule and
+  // it works similarly to "SpokenAttributeToDLIR".
+  DLIRAssertion SpokenCanActAsToDLIR(const Principal& speaker,
+                                     const CanActAs& can_act_as);
+
+  std::pair<Predicate, std::vector<DLIRAssertion>> BaseFactToDLIRInner(
+      const Principal& speaker, const Predicate& predicate) {
+    std::vector<DLIRAssertion> pred_list = {};
+    return std::make_pair(predicate, pred_list);
+  }
+
+  std::pair<Predicate, std::vector<DLIRAssertion>> BaseFactToDLIRInner(
+      const Principal& speaker, const Attribute& attribute) {
+    std::vector<DLIRAssertion> pred_list = {
+        SpokenAttributeToDLIR(speaker, attribute)};
+    return std::make_pair(AttributeToDLIR(attribute), pred_list);
```

Comment:
Use `std::move(pred_list)` to avoid a copy. 

Alternatively, the following should avoid a copy too:
```
    return std::make_pair(AttributeToDLIR(attribute), {SpokenAttributeToDLIR(...)}); 
```
However, it might make the code less readable.